src/main/scala/com/microsoft/hyperspace/util/ResolverUtils.scala

```diff
@@ -16,19 +16,73 @@
 
 package com.microsoft.hyperspace.util
 
-import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.{Column, SparkSession}
 import org.apache.spark.sql.catalyst.analysis.Resolver
 import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, Expression, ExtractValue, GetArrayStructFields, GetMapValue, GetStructField}
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.functions.col
 import org.apache.spark.sql.types.{ArrayType, MapType, StructField, StructType}
 
 import com.microsoft.hyperspace.HyperspaceException
+import com.microsoft.hyperspace.util.ResolverUtils.ResolvedColumn.NESTED_FIELD_PREFIX
 
 /**
  * [[ResolverUtils]] provides utility functions to resolve strings based on spark's resolver.
  */
 object ResolverUtils {
 
+  /**
+   * [[ResolvedColumn]] stores information when a column name is resolved against the
+   * analyzed plan and its schema.
+   *
+   * Outside unit tests, this object should not be created directly, but via the `resolve` function,
+   * or `ResolvedColumn.apply` with a normalized name.
+   *
+   * @param name The column name resolved from an analyzed plan.
+   * @param isNested Flag to denote if this column is nested or not.
+   */
+  private[hyperspace] case class ResolvedColumn(name: String, isNested: Boolean) {
+    assert(!isNested || (name.contains(".") && !name.startsWith(NESTED_FIELD_PREFIX)))
+
+    // For nested fields, the column name is prefixed with `NESTED_FIELD_PREFIX`.
+    lazy val normalizedName = {
+      if (isNested) {
+        s"$NESTED_FIELD_PREFIX$name"
+      } else {
+        name
+      }
+    }
+
+    // Create a column using the resolved name. Top level column names are quoted, and
+    // nested column names are aliased with normalized names.
+    def toColumn: Column = {
+      if (isNested) {
+        // No need to quote the string for "as" even if it contains dots.
+        col(name).as(normalizedName)
+      } else {
+        col(quote(name))
+      }
+    }
+
+    // Create a column using the normalized name. Since the normalized name is already flattened
+    // with "dots", it is quoted.
+    def toNormalizedColumn: Column = col(quote(normalizedName))
+
+    private def quote(name: String) = s"`$name`"
```

Comment:
Can there be a case where a nested field access path contains quotes and become twice quoted? 

For example, given the following dataset schema:
```
root
 |-- id: integer (nullable = true)
 |-- nest: struct (nullable = true)
 |    |-- n.dot.nested: struct (nullable = true)
 |    |    |-- field: string (nullable = true)
```
In Spark it's correct to access `field` with this construct: ``nest.`n.dot.nested`.field``. 

Can we get into the case of quoting the ``nest.`n.dot.nested`.field`` ending up with `` `nest.`n.dot.nested`.field` ``? Should we check if field does already contain backticks?

Could you add some unit tests for this use case?

Reply:
As the following comment says, the `name` passed to the `ResolvedColumn` should be the result of `resolve`: "Outside unit tests, this object should not be created directly, but via the `resolve` function". So, it shouldn't receive the example you posted. But, we could add more asserts here though.